sdk/eventhub/event-hubs/src/util/connectionStringUtils.ts

```diff
@@ -0,0 +1,139 @@
+  /**
+   * The value for "EntityPath" in the connection string which would be the name of the event hub instance associated with the connection string.
+   * Connection string from a Shared Access Policy created at the namespace level
+   * will not have the EntityPath in it.
+   */
+  eventHubName?: string;
+  /**
+   * The value for "SharedAccessKey" in the connection string. This along with the "SharedAccessKeyName"
+   * in the connection string is used to generate a SharedAccessSignature which can be used authorize
+   * the connection to the service.
+   */
+  sharedAccessKey?: string;
+  /**
+   * The value for "SharedAccessKeyName" in the connection string. This along with the "SharedAccessKey"
+   * in the connection string is used to generate a SharedAccessSignature which can be used authorize
+   * the connection to the service.
+   */
+  sharedAccessKeyName?: string;
+  /**
+   * The value for "SharedAccessSignature" in the connection string. This is typically not present in the
+   * connection string generated for a Shared Access Policy. It is instead generated by the
+   * user and appended to the connection string for ease of use.
+   */
+  sharedAccessSignature?: string;
+}
+
+/**
+ * Parses given connection string into the different properties applicable to Azure Event Hubs.
+ * The properties are useful to then construct an EventHub{Producer|Consumer}Client.
+ * @param connectionString The connection string associated with the Shared Access Policy created
+ * for the Event Hubs namespace.
+ */
+export function parseEventHubConnectionString(
+  connectionString: string
+): Readonly<EventHubConnectionStringProperties> {
+  const parsedResult = parseConnectionString<GenericConnectionStringProperties>(connectionString);
+
+  const validationResult = validate(parsedResult);
+
+  if (!validationResult.isValid) {
+    throw new Error(validationResult.message);
+  }
+
+  const output: EventHubConnectionStringProperties = {
+    fullyQualifiedNamespace: (parsedResult.Endpoint.match(".*://([^/]*)") || [])[1],
+    endpoint: parsedResult.Endpoint
+  };
+
+  if (parsedResult.EntityPath) {
+    output.eventHubName = parsedResult.EntityPath;
+  }
+
+  if (parsedResult.SharedAccessSignature) {
+    output.sharedAccessSignature = parsedResult.SharedAccessSignature;
+  }
+
+  if (parsedResult.SharedAccessKey && parsedResult.SharedAccessKeyName) {
+    output.sharedAccessKey = parsedResult.SharedAccessKey;
+    output.sharedAccessKeyName = parsedResult.SharedAccessKeyName;
+  }
+
+  return output;
+}
+
+/**
+ * @internal
+ * @ignore
+ */
+type GenericConnectionStringProperties = {
+  Endpoint: string;
+  EntityPath?: string;
+  SharedAccessSignature?: string;
+  SharedAccessKey?: string;
+  SharedAccessKeyName?: string;
+};
+
+/**
+ * @internal
+ * @ignore
+ */
+type ConnectionStringValidationResult =
+  | { isValid: true }
+  | {
+      isValid: false;
+      message: string;
+    };
+
+/**
+ * @internal
+ * @ignore
+ */
+function validate(
```

Comment:
Just curious... What made us pull out the validation logic and introduce a new function and two new types? I see this approach being very useful when we want to re-use the code in this function and types or even unit test the function, but from what I can see these are used only in 1 place i.e. in the `parseEventHubConnectionString` method

Reply:
I think I feel good about my decision to pull out the validation logic to its own function as it helps with readability here -at least in my opinion it's nice to be able to focus just on what makes a connection string valid vs what we map it to. Of course it's totally opinion based here especially since it's not used anywhere else.

Looking at this code again I think that the new type is probably not necessary but I'm curious if there's harm in introducing it and if you have a recommendation for what `validate` should return instead? It can return an anonymous type, an untyped object, or I can inline the `validate` method logic (but I'd prefer to keep the two separate if possible).

Reply:
> I think I feel good about my decision to pull out the validation logic to its own function as it helps with readability here -at least in my opinion it's nice to be able to focus just on what makes a connection string valid vs what we map it to.

Yes, readability is key. I always feel conflicted between this approach and the ideology that adding a new function introduces a level of indirection. I prefer the modular approach if the logic gets used in multiple places. Thus the question. In this case, we should be good. The function name and signature is intuitive enough to not cause the reader any distraction from the indirection. 

> Looking at this code again I think that the new type is probably not necessary but I'm curious if there's harm in introducing it?

As a reader/maintainer of the code, it is yet another type to understand. So, unless it brings in value in terms of re-usability, I tend to shy away from introducing new types.

> if you have a recommendation for what validate should return instead

What about throwing the error from the validate() function itself? Then it need not return anything.